File: vol2/ch22/LeetCode1584-1.py
class Solution:
  def minCostConnectPoints(self, points: List[List[int]]) -> int:
    n=len(points)
    if n==1:return 0
    if n==2:return self.dist(points,0,1)
    return self.Prim(points,0)
  def dist(self,points,i,j):  #求曼哈顿距离
    return abs(points[i][0]-points[j][0])+abs(points[i][1]-points[j][1])
  def Prim(self,points,v):        #Prim算法
    INF=0x3f3f3f3f
    n=len(points)
    lowcost,closest=[0]*n,[0]*n
    ans=0
    for i in range(0,n):  #给lowcost[]和closest[]置初值
      lowcost[i]=self.dist(points,v,i)
      closest[i]=v
    for i in range(1,n):  #找出(n-1)个顶点
      mind,k=INF,-1
      for j in range(0,n):  #在(V-U)中找出离U最近的顶点k
        if lowcost[j]!=0 and lowcost[j]<mind:
          mind=lowcost[j]
          k=j              #k记录当前最小边的顶点
      ans+=mind           #生成最小生成树的一条边
      lowcost[k]=0         #标记k已经加入U
      for j in range(0,n):     #对(V-U)中顶点j进行调整
        if lowcost[j]!=0 and self.dist(points,k,j)<lowcost[j]:
          lowcost[j]=self.dist(points,k,j)
          closest[j]=k      #修改数组lowcost和closest
    return ans

File: vol2/ch22/LeetCode1584-1.cpp
#include<vector>
#include<iostream>
#include<queue>
using namespace std;

struct Edge {                                     //出边类
    int vno;                                      //邻接点
    int wt;                                       //边的权
    Edge(int v,int w):vno(v),wt(w) {}
};
struct QNode {                  //优先队列结点类型
    int vno;                    //顶点编号
    int length;                 //路径长度
    int nums;                   //中转的城市数
    bool operator<(const QNode& b) const {
        return length>b.length;     //length越小越优先出队
    }
};
class Solution {
    const int INF=0x3f3f3f3f;                                   //表示∞
public:
    int findCheapestPrice(int n,vector<vector<int>>&flights,int src,int dst,int k) {
        vector<vector<Edge>> adj(n);            //邻接表
        for(int i=0;i<flights.size();i++) {   //遍历flights建立邻接表
            int a=flights[i][0];
            int b=flights[i][1];
            int w=flights[i][2];                //<a,b>:w
            adj[a].push_back(Edge(b,w));
        }
        vector<int> dist(n,INF);								//最短路径长度
        vector<int> cnt(n,0);									//最小中转次数
        priority_queue<QNode> pqu;               //小根堆
        QNode e,e1,e2;
        e.vno=src;e.length=0;e.nums=0;
        pqu.push(e);                     //源点元素进队
        dist[src]=0;
        cnt[src]=0;
        while(!pqu.empty()) {
            auto e=pqu.top();pqu.pop();
            int u=e.vno,length=e.length,nums=e.nums;
            if(u==dst) return length;
            if(nums>k) continue;            //剪支
            for(auto edj:adj[u]) {				//扩展
                int v=edj.vno,w=edj.wt;
                if(length+w<dist[v]) {			//剪支：边松驰（合并）
                    dist[v]=length+w;
                    cnt[v]=nums+1;
                    e1.vno=v;e1.length=dist[v];e1.nums=cnt[v];
                    pqu.push(e1);
                }
                else if(nums+1<cnt[v]) {		//尽管路径长度不是更短，但中转次数较少
                    e2.vno=v;e2.length=length+w;
                    e2.nums=nums+1;
                    pqu.push(e2);
                }
            }
        }
        return -1;
    }
};

